File: frontend/console/inventory/src/application/listCreate/listCreate.dto.ts
// frontend/console/inventory/src/application/listCreate/listCreate.dto.ts

import type { ListCreateDTO } from "../../infrastructure/http/inventoryRepositoryHTTP";
import type { PriceRow } from "./listCreate.types";

export function extractDisplayStrings(dto: ListCreateDTO | null): {
  productBrandName: string;
  productName: string;
  tokenBrandName: string;
  tokenName: string;
} {
  return {
    productBrandName: dto?.productBrandName ?? "",
    productName: dto?.productName ?? "",
    tokenBrandName: dto?.tokenBrandName ?? "",
    tokenName: dto?.tokenName ?? "",
  };
}

/**
 * backend の ListCreateDTO.priceRows を PriceCard 用 PriceRow[] に変換
 * - 期待値: inventory/application の PriceRow（listCreate.types.ts）を正とする
 * - 識別子は modelId を正とする
 * - 並び順は displayOrder（未設定は null を保持）
 * - 名揺れ補完はしない
 */
export function mapDTOToPriceRows(dto: ListCreateDTO | null): PriceRow[] {
  const rows = Array.isArray(dto?.priceRows) ? dto.priceRows : [];

  return rows.map((r: any) => {
    const displayOrderRaw = r.displayOrder;
    const displayOrder =
      displayOrderRaw === null || displayOrderRaw === undefined
        ? null
        : Number(displayOrderRaw);

    const stockRaw = Number(r.stock ?? 0);
    const stock = Number.isFinite(stockRaw) ? stockRaw : 0;

    const row: PriceRow = {
      modelId: r.modelId,
      displayOrder,
      size: r.size,
      color: r.color,
      stock,
      rgb: r.rgb as any,
      price: r.price === undefined ? null : (r.price as any),
    };

    return row;
  });
}

/**
 * 初期表示用: PriceRow[] を返す
 */
export function initPriceRowsFromDTO(dto: ListCreateDTO | null): PriceRow[] {
  return mapDTOToPriceRows(dto);
}

File: backend/internal/application/query/console/dto/list_create_dto.go
// backend/internal/application/query/console/dto/list_create_dto.go
package dto

// ListCreateDTO is a minimal DTO for listCreate screen.
// Current requirements:
// - From ProductBlueprint: brandName, productName
// - From TokenBlueprint: tokenName, brandName
// - PriceCard: priceRows
// - Transportation: yamato, sagawa, post, custom
type ListCreateDTO struct {
	// From ProductBlueprint
	ProductBrandName string `json:"productBrandName"`
	ProductName      string `json:"productName"`

	// From TokenBlueprint
	TokenBrandName string `json:"tokenBrandName"`
	TokenName      string `json:"tokenName"`

	// PriceCard 用
	PriceRows []ListCreatePriceRowDTO `json:"priceRows,omitempty"`

	// 配送方法・配送料金設定
	TransportationOptions []ListCreateTransportationOptionDTO `json:"transportationOptions,omitempty"`
}

// ListCreateTransportationOptionDTO is a transportation option for listCreate screen.
//
// TransportationOption:
// - yamato
// - sagawa
// - post
// - custom
//
// TransportationID:
// - custom の場合のみ設定する。
// - backend/internal/domain/transportation.TransportationFeeSetting.ID を保持する。
// - yamato / sagawa / post の場合は空文字とする。
type ListCreateTransportationOptionDTO struct {
	TransportationOption string `json:"transportationOption"`
	TransportationID     string `json:"transportationId,omitempty"`
	Name                 string `json:"name"`
}

// ListCreateModelRefDTO is a lightweight ModelRef for query internal use.
// - displayOrder は「取得するのみ」
// - 0/未設定は null 扱いに寄せる
type ListCreateModelRefDTO struct {
	ModelID      string `json:"modelId"`
	DisplayOrder *int   `json:"displayOrder,omitempty"`
}

// ListCreatePriceRowDTO is a row DTO for PriceCard.
// - 更新や作成 payload で識別できるよう ModelID は保持する。
// - productBlueprintCategory / model kind に応じた model 情報を含める。
type ListCreatePriceRowDTO struct {
	ModelID string `json:"modelId"`

	// model kind
	// - apparel
	// - alcohol
	Kind string `json:"kind,omitempty"`

	// displayOrder（ProductBlueprintPatch.ModelRefs.DisplayOrder）
	// - 取得するのみ（サーバ側で並べ替えしない）
	// - 0/未設定は null 扱いに寄せる
	DisplayOrder *int `json:"displayOrder,omitempty"`

	// 在庫数
	Stock int `json:"stock"`

	// apparel 系表示用
	Size  string `json:"size,omitempty"`
	Color string `json:"color,omitempty"`
	RGB   *int   `json:"rgb,omitempty"`

	// alcohol 系表示用
	VolumeValue *int   `json:"volumeValue,omitempty"`
	VolumeUnit  string `json:"volumeUnit,omitempty"`

	// 価格
	// - 未入力: null
	Price *int `json:"price,omitempty"`
}
